flocker/node/_docker.py

```diff
@@ -752,39 +752,51 @@ def _remove():
+                    # 404 error, even though the container exists.
+                    # See https://github.com/docker/docker/issues/13088
+                    # Wait until the container has actually stopped running
+                    # before attempting to remove it.  Otherwise we are
+                    # likely to see: 'docker.errors.APIError: 409 Client
+                    # Error: Conflict ("Conflict, You cannot remove a
+                    # running container. Stop the container before
+                    # attempting removal or use -f")'
+                    # This code should probably be removed once the above
+                    # issue has been resolved. See [FLOC-1850]
+                    self._client.wait(container_name)
 
-                Message.new(
-                    message_type="flocker:docker:container_remove",
-                    container=container_name
-                ).write()
-                self._client.remove_container(container_name)
-                Message.new(
-                    message_type="flocker:docker:container_removed",
-                    container=container_name
-                ).write()
-            except APIError as e:
-                # If the container doesn't exist, we swallow the error,
-                # since this method is supposed to be idempotent.
-                if e.response.status_code == NOT_FOUND:
                     Message.new(
-                        message_type="flocker:docker:container_not_found",
+                        message_type="flocker:docker:container_remove",
                         container=container_name
                     ).write()
-                    return
-                raise
+                    self._client.remove_container(container_name)
+                    Message.new(
+                        message_type="flocker:docker:container_removed",
+                        container=container_name
+                    ).write()
+                except APIError as e:
+                    # If the container doesn't exist, we swallow the error,
+                    # since this method is supposed to be idempotent.
+                    if e.response.status_code == NOT_FOUND:
+                        Message.new(
+                            message_type="flocker:docker:container_not_found",
+                            container=container_name
+                        ).write()
+                        return
+                    # Sometimes Docker returns a 500 error when trying to
+                    # remove a container which it is already removing.
+                    # This has been resolved in https://github.com/docker/docker/commit/c4e49d10143178c718178ff7ce857f4f8ed46a0b  # noqa
+                    # This code should be removed once we drop support for
+                    # Docker versions older than 1.9.0. See FLOC-3284.
+                    if e.response.status_code == INTERNAL_SERVER_ERROR:
```

Comment:
I'm interested in hearing review feedback on whether this should have a more specific check, e.g. of the response message